commit e0ae1232255edaae5f93789c12528c5da8097a23
Author: Eike Hein <hein@kde.org>
Date:   Fri Mar 21 08:01:46 2014 +0100

    Don't show dupes in All Apps.
    
    BUG:327026

diff --git a/components/sources/installedapps/installedappsmodel.cpp b/components/sources/installedapps/installedappsmodel.cpp
index 6069a70..9215f0e 100644
--- a/components/sources/installedapps/installedappsmodel.cpp
+++ b/components/sources/installedapps/installedappsmodel.cpp
@@ -394,7 +394,19 @@ void InstalledAppsModel::doLoadServiceGroup(KServiceGroup::Ptr group)
                 if (genericName.isNull()) {
                     genericName = service->comment();
                 }
-                m_nodeList << new AppNode(service, this);
+
+                bool found = false;
+
+                foreach(const AbstractNode *node, m_nodeList) {
+                    if (node->type() == AbstractNode::AppNodeType
+                        && static_cast<const AppNode *>(node)->service()->storageId() == service->storageId()) {
+                        found = true;
+                    }
+                }
+
+                if (!found) {
+                    m_nodeList << new AppNode(service, this);
+                }
             }
 
         } else if (p->isType(KST_KServiceGroup)) {
